### PR TITLE
DM-23795: Modernize packaging and add type checking

### PR DESCRIPTION
**Pull request by jonathansick**
We've significantly improved Kafkit's packaging and infrastructure:

- Migrate packaging metadata from ``setup.py`` to ``setup.cfg``.
- Support PEP 518 with ``pyproject.toml`` (Kafkit continues to be a setuptools-based project).
- Tox now runs tasks like tests, in conjunction with the existing Pytest set up.
- Pre-commit hooks lint and format the code base.
- Code style is now handled by Black (and in the documentation with blacken-docs).
- **Full support for type annotations!** ``tox -e typing`` validates Kafkit's type annotations with Mypy.
- We've migrated from Travis CI to GitHub Actions.
- The documentation now includes a development guide.

### CODE REVIEW

File: src/kafkit/registry/sansio.py [hunk 1]
```diff
@@ -184,7 +212,9 @@ def subject_cache(self):
         data = await self._make_request("GET", url, url_vars, b"")
         return data
 
-    async def post(self, url, url_vars=dict(), *, data):
+    async def post(
+        self, url: str, url_vars: Mapping[str, str] = dict(), *, data: Any
```
Comment (rra) [inline src/kafkit/registry/sansio.py]: Same issue here.

File: src/kafkit/registry/sansio.py [hunk 2]
```diff
@@ -254,7 +286,9 @@ def subject_cache(self):
         data = await self._make_request("PATCH", url, url_vars, data)
         return data
 
-    async def put(self, url, url_vars=dict(), data=b""):
+    async def put(
+        self, url: str, url_vars: Mapping[str, str] = dict(), data: Any = b""
```
Comment (rra) [inline src/kafkit/registry/sansio.py]: Same issue here.

File: src/kafkit/registry/sansio.py [hunk 3]
```diff
@@ -289,7 +323,13 @@ def subject_cache(self):
         data = await self._make_request("PUT", url, url_vars, data)
         return data
 
-    async def delete(self, url, url_vars=dict(), *, data=b""):
+    async def delete(
+        self,
+        url: str,
+        url_vars: Mapping[str, str] = dict(),
```
Comment (rra) [inline src/kafkit/registry/sansio.py]: Same issue here.

File: src/kafkit/registry/sansio.py [hunk 4]
```diff
@@ -511,22 +560,29 @@ class MockRegistryApi(RegistryApi):
...
-    async def _request(self, method, url, headers, body):
+    async def _request(
+        self, method: str, url: str, headers: Mapping[str, str], body: bytes
+    ) -> Any:
         self.method = method
         self.url = url
         self.headers = headers
         self.body = body
-        response_headers = self.response_headers.copy()
+        response_headers = dict(self.response_headers)  # make a copy
```
Comment (rra) [inline src/kafkit/registry/sansio.py]: Very minor nit: I tend to use `copy.deepcopy` in cases like this just to make the copy explicit.

File: src/kafkit/registry/serializer.py [hunk 5]
```diff
@@ -162,13 +185,17 @@ def __init__(self, *, registry):
             schema = await self._registry.get_schema_by_id(schema_id)
         elif schema is not None:
             schema_id = await self._registry.register_schema(
-                schema=schema, subject=subject)
+                schema=schema, subject=subject
+            )
         else:
-            raise RuntimeError('Pass either a schema or schema_id parameter.')
+            raise RuntimeError("Pass either a schema or schema_id parameter.")
+        assert isinstance(schema, dict)  # to guide mypy
```
Comment (rra) [inline src/kafkit/registry/serializer.py]: In this case, you're just unconfusing mypy about whether `schema` is `None`, so I would simplify this to `assert schema`. The `isinstance` makes me think something more complicated is going on than it actually is.

Another option that avoids the `assert` (but which may seem awkward; it's a matter of taste) to replace the `else` clause with:
```
if not schema:
    raise RuntimeError("Pass either a schema or schema_id parameter")
```

**Comment by jonathansick** [comment]
Great point. Thanks for taking a look.

**Pull request merged**